packages/react/src/popover/viewport/PopoverViewport.tsx

```diff
@@ -0,0 +1,219 @@
+import * as React from 'react';
+import { useStore } from '@base-ui-components/utils/store';
+import { useAnimationFrame } from '@base-ui-components/utils/useAnimationFrame';
+import { usePrevious } from '@base-ui-components/utils/usePrevious';
+import { useIsoLayoutEffect } from '@base-ui-components/utils/useIsoLayoutEffect';
+import { usePopoverRootContext } from '../root/PopoverRootContext';
+import { BaseUIComponentProps } from '../../utils/types';
+import { useAnimationsFinished } from '../../utils/useAnimationsFinished';
+import { useRenderElement } from '../../utils/useRenderElement';
+import { CustomStyleHookMapping } from '../../utils/getStyleHookProps';
+import { selectors } from '../store';
+
+const customStyleHookMapping: CustomStyleHookMapping<PopoverViewport.State> = {
+  activationDirection: (value) =>
+    value
+      ? {
+          'data-activation-direction': value,
+        }
+      : null,
+};
+
+/**
+ * A viewport for displaying content transitions.
+ * This component is only required if one popup can be opened by multiple triggers, its content change based on the trigger
+ * and switching between them is animated.
+ * Renders a `<div>` element.
+ *
+ * Documentation: [Base UI Popover](https://base-ui.com/react/components/popover)
+ */
+export const PopoverViewport = React.forwardRef(function PopoverViewport(
+  componentProps: PopoverViewport.Props,
+  forwardedRef: React.ForwardedRef<HTMLDivElement>,
+) {
+  const { render, className, children, ...elementProps } = componentProps;
+  const { store } = usePopoverRootContext();
+
+  const activeTrigger = useStore(store, selectors.activeTriggerElement);
+  const previousActiveTrigger = usePrevious(activeTrigger);
+
+  const capturedHtmlRef = React.useRef<string | null>(null);
+  const [previousChildrenHtml, setPreviousChildrenHtml] = React.useState<string | null>(null);
+
+  const [newTriggerOffset, setNewTriggerOffset] = React.useState<Offset | null>(null);
+  const currentContentRef = React.useRef<HTMLDivElement>(null);
+  const nextContainerRef = React.useRef<HTMLDivElement>(null);
+  const onAnimationsFinished = useAnimationsFinished(nextContainerRef, true);
+
+  const cleanupTimeout = useAnimationFrame();
+
+  const [previousContentDimensions, setPreviousContentDimensions] = React.useState<{
+    width: number;
+    height: number;
+  } | null>(null);
+
+  // Capture children HTML in a ref when not transitioning.
+  // We can't simply store previous `children` React node, as it might be stateful and doing so would lose this state.
+  useIsoLayoutEffect(() => {
+    if (currentContentRef.current && !previousChildrenHtml) {
+      capturedHtmlRef.current = currentContentRef.current.innerHTML;
+    }
+  });
```

Comment:
`innerHTML` can be dangerous 😬  is there a way to avoid this?

Reply:
I replaced this code with capturing the actual DOM node and cloning it into the `previous` container. Should also be faster than serializing and deserializing to HTML.